### PR TITLE
feat: add conversion of leaky relu from onnx to tosa

### PR DESCRIPTION
**Pull request by maxbartel**
Most Functions from the Utils files are copied either from the TF or torch-mlir repo. They too share the same functions.

### CODE REVIEW

**Review by ljfitz — CHANGES_REQUESTED** [review]
Looking pretty good. There are some details I'd like you to consider though.

File: src/Conversion/ONNXToTOSA/ONNXToTOSALegalizeUtils.hpp [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+#include "mlir/IR/BuiltinTypes.h"                 // from @llvm-project
+#include "mlir/IR/PatternMatch.h"                 // from @llvm-project
+#include "mlir/Interfaces/InferTypeOpInterface.h" // from @llvm-project
+#include "mlir/Support/LLVM.h"                    // from @llvm-project
+
+namespace mlir {
+namespace tosa {
+
+// Create a 32-bit float constant operator from a float
+Value getTosaConstTensorSingleF32(PatternRewriter &rewriter, Operation *op,
```
Comment (ljfitz) [inline src/Conversion/ONNXToTOSA/ONNXToTOSALegalizeUtils.hpp:32]: As this differs from `tensorflow` in the `shape` argument, this needs to be documented too.

File: src/Conversion/ONNXToTOSA/Math/Elementwise.cpp [hunk 2]
```diff
@@ -79,13 +80,58 @@ class ONNXReluOpLoweringToTOSA : public OpConversionPattern<ONNXReluOp> {
...
+
+    TensorType outputType = op.getResult().getType().dyn_cast<TensorType>();
+
+    if (!outputType.getElementType().isF32()) {
+      return rewriter.notifyMatchFailure(op, "Only float is supported");
+    }
+
+    FloatAttr alphaAttr = adaptor.alphaAttr();
+
+    double alpha = 0.01;
```
[the remark this answers is not in the document]
Comment (maxbartel) [reply src/Conversion/ONNXToTOSA/Math/Elementwise.cpp:99]: There is actually no equivalent to getValueAsDouble for float. To create a F32 Tensor we need to cast it at one point.

File: src/Conversion/ONNXToTOSA/ONNXToTOSALegalizeUtils.hpp (hunk 1 above)
Comment (maxbartel) [reply src/Conversion/ONNXToTOSA/ONNXToTOSALegalizeUtils.hpp:35]: Added the documentation

File: src/Conversion/ONNXToTOSA/ONNXToTOSALegalizeUtils.cpp [hunk 3]
```diff
@@ -0,0 +1,54 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//====------ ONNXToTOSACommon.hpp - ONNX dialects to TOSA lowering Utils---===//
+//
+// Copyright (c) 2022 Advanced Micro Devices, Inc.
```
[the remark this answers is not in the document]
Comment (ljfitz) [reply src/Conversion/ONNXToTOSA/ONNXToTOSALegalizeUtils.cpp:8]: @luis - could you have a look and see if you consider the copyright handling to be okay:
- this file copies some code from https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/mlir/tosa/transforms/legalize_utils.cc
- Max has added some code, and made some modifications to others
- There is a goal (in the future) to have common code moved to a common place for tensorflow+torch+onnx to tosa

File: src/Conversion/ONNXToTOSA/Math/Elementwise.cpp [hunk 4]
```diff
@@ -79,13 +80,58 @@ class ONNXReluOpLoweringToTOSA : public OpConversionPattern<ONNXReluOp> {
...
+
+    if (alphaAttr) {
+      alpha = alphaAttr.getValueAsDouble();
+    }
```
[the remark this answers is not in the document]
Comment (maxbartel) [reply src/Conversion/ONNXToTOSA/Math/Elementwise.cpp]: Fixed, alpha is now float

**Review by ljfitz — APPROVED** [review]
I had some comments in `pending` and added one suggested reworking around `reduceAxisToOne...`. Up to you to decide if they are worthwhile.

File: src/Conversion/ONNXToTOSA/ONNXToTOSALegalizeUtils.cpp [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    mlir::Attribute encoding = {}) {
+  return mlir::RankedTensorType::get(
+      llvm::SmallVector<int64_t, 4>(shape.size(), 1), elementType, encoding);
+}
+
+// Create a 32-bit float constant operator from a float
+Value getTosaConstTensorSingleF32(PatternRewriter &rewriter, Operation *op,
+    float val, llvm::ArrayRef<int64_t> shape) {
+  auto constType =
+      reduceAxisToOneAndGetTypeFromTensorShape(shape, rewriter.getF32Type());
```
[the remark this answers is not in the document]
Comment (maxbartel) [reply src/Conversion/ONNXToTOSA/ONNXToTOSALegalizeUtils.cpp]: Yours is better, I will adapt that

**Review by flemairen6 — APPROVED** [review]

**Pull request merged**